### PR TITLE
enable support of xpubs for bitcoin cash

### PR DESCRIPTION
**Pull request by prettyirrelevant**

### CODE REVIEW

**Review by yabirgb — COMMENTED** [review]
Small comments

File: rotkehlchen/tests/db/test_db_upgrades.py [hunk 1]
```diff
@@ -2509,6 +2510,14 @@ def test_upgrade_db_31_to_32(user_data_dir):  # pylint: disable=unused-argument
     cursor.execute('SELECT COUNT(*), timestamp FROM history_events WHERE subtype="staking receive asset" AND type="unstaking"')  # noqa: E501
     assert cursor.fetchone() == (1, expected_timestamp)
 
+    # check that you cannot add blockchain column in xpub_mappings
```
Comment (LefterisJP) [inline rotkehlchen/tests/db/test_db_upgrades.py:2513]: no no no absolutely not. As discussed you **can not** modify a DB upgrade that has already been released to the public. Make a new one and test it properly. You will also need to make a new empty test v32 DB and populate it with xpubs and see they get upgraded properly.

**Review by LefterisJP — COMMENTED** [review]
Hmmm lots of comments.

The biggest things to consider are:

1. Please do not duplicate code. Abstract common code away into functions. BCH/BTC xpub derivation code is identical
2. **Never ever** modify a DB upgrade that has gone live to add new schema modifications. The only reason to do that is for a bugfix.

File: rotkehlchen/db/dbhandler.py [hunk 2]
```diff
@@ -3215,7 +3215,11 @@ def add_bitcoin_xpub(self, xpub_data: XpubData) -> None:
             ) from e
         self.update_last_write()
 
-    def delete_bitcoin_xpub(self, xpub_data: XpubData) -> None:
+    def delete_bitcoin_xpub(
```
[the remark this answers is not in the document]
Comment (prettyirrelevant) [reply rotkehlchen/db/dbhandler.py:3218]: The `add_bitcoin_xpub` function only creates an xpub entry using the following `(xpub, derivation_path, label)`. 
The blockchain parameter is only needed when creating an entry in `xpub_mappings`

File: rotkehlchen/tests/api/test_bitcoin.py [hunk 3]
```diff
@@ -242,6 +242,43 @@ def test_add_delete_xpub(rotkehlchen_api_server):
...
+    }
+    response = requests.put(api_url_for(
+        rotkehlchen_api_server,
+        "bchxpubresource",
+    ), json=json_data)
+    if async_query:
+        task_id = assert_ok_async_response(response)
+        outcome = wait_for_async_task_with_result(rotkehlchen_api_server, task_id, timeout=180)
+    else:
+        outcome = assert_proper_response_with_result(response)
```
[the remark this answers is not in the document]
Comment (prettyirrelevant) [reply rotkehlchen/tests/api/test_bitcoin.py:260]: Unless I don't get you but `wait_for_async_task_with_result` uses `wait_for_async_task` underneath. The only extra thing it does is to return the result.

File: rotkehlchen/chain/bitcoin/bch/__init__.py [hunk 4]
```diff
@@ -47,13 +48,66 @@ def get_bitcoin_cash_addresses_balances(accounts: List[BTCAddress]) -> Dict[BTCA
...
+    Takes a list of BCH addresses and returns a mapping of which addresses have had transactions
+    and also their current balance
+
+    May raise:
+    - RemoteError if any of the queried websites fail to be queried
+    """
+    try:
+        have_transactions = _check_haskoin_for_transactions(accounts)
+    except requests.exceptions.RequestException as e:
+        raise RemoteError(f'bitcoin external API request for transactions failed due to {str(e)}') from e  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/chain/bitcoin/bch/__init__.py:102]:
```suggestion
        raise RemoteError(f'bitcoin cash external API request for transactions failed due to {str(e)}') from e  # noqa: E501
```

File: rotkehlchen/chain/bitcoin/bch/__init__.py [hunk 5]
```diff
@@ -47,13 +48,66 @@ def get_bitcoin_cash_addresses_balances(accounts: List[BTCAddress]) -> Dict[BTCA
...
+    except KeyError as e:
+        raise RemoteError(
+            f'Malformed response when querying bitcoin blockchain via {HASKOIN_BASE_URL}. '
```
Comment (LefterisJP) [inline rotkehlchen/chain/bitcoin/bch/__init__.py:105]:
```suggestion
            f'Malformed response when querying BCH blockchain via {HASKOIN_BASE_URL}. '
```

File: rotkehlchen/chain/bitcoin/bch/__init__.py [hunk 6]
```diff
@@ -47,13 +48,66 @@ def get_bitcoin_cash_addresses_balances(accounts: List[BTCAddress]) -> Dict[BTCA
...
+            f'Did not find key {str(e)}',
+        ) from e
+    except DeserializationError as e:
+        raise RemoteError(
+            f'Malformed response when querying Bitcoin Cash blockchain via {HASKOIN_BASE_URL}. '
```
Comment (LefterisJP) [inline rotkehlchen/chain/bitcoin/bch/__init__.py:110]: just for uniformity
```suggestion
            f'Malformed response when querying BCH blockchain via {HASKOIN_BASE_URL}. '
```

File: rotkehlchen/chain/manager.py [hunk 7]
```diff
@@ -312,6 +294,46 @@ def account_exists(
         # else
         raise AssertionError('Invalid blockchain value')
 
+    def _serialize_bitcoin_balances(
+        self,
+        bitcoin_balances: Dict[str, Any],
+        xpub_mappings: Dict[BTCAddress, XpubData],
+        blockchain: Literal[SupportedBlockchain.BITCOIN, SupportedBlockchain.BITCOIN_CASH],
+    ) -> None:
+        """This is an helper function to serialize the balances for BTC & BCH accounts."""
```
Comment (LefterisJP) [inline rotkehlchen/chain/manager.py:303]:
```suggestion
        """This is a helper function to serialize the balances for BTC & BCH accounts."""
```

File: rotkehlchen/chain/manager.py [hunk 8]
```diff
@@ -771,23 +793,33 @@ def query_polkadot_balances(
             total_balance += balance
         self.totals.assets[A_DOT] = total_balance
 
-    def sync_btc_accounts_with_db(self) -> None:
-        """Call this function after having deleted BTC accounts from the DB to
+    def sync_btc_accounts_with_db(
```
Comment (LefterisJP) [inline rotkehlchen/chain/manager.py:796]: rename to make it obvious it's for all bitcoin chains.

```suggestion
    def sync_bitcoin_accounts_with_db(
```

File: rotkehlchen/chain/manager.py [hunk 9]
```diff
@@ -771,23 +793,33 @@ def query_polkadot_balances(
...
+        for account in getattr(self.accounts, blockchain.value.lower()):
+            if account not in db_btc_accounts:
+                accounts_to_remove.append(account)
 
-        balances_mapping = get_bitcoin_addresses_balances(accounts_to_remove)
+        get_blockchain_balances = {
+            'BCH': get_bitcoin_cash_addresses_balances,
+            'BTC': get_bitcoin_addresses_balances,
+        }
+        balances_mapping = get_blockchain_balances[blockchain.value](accounts_to_remove)
```
Comment (LefterisJP) [inline rotkehlchen/chain/manager.py:819]: Hmmm what are you doing here?

Why are you calling both `get_bitcoin_cash_addresses_balances` and `get_bitcoin_addresses_balances`?

In the end in line 819 you only use one of those. Why go to the extra trouble and not just the appropriate one depending on the passed `blockchain` value?

File: rotkehlchen/tests/db/test_db_upgrades.py [hunk 10]
```diff
@@ -2602,6 +2603,45 @@ def test_upgrade_db_31_to_32(user_data_dir):  # pylint: disable=unused-argument
...
+    )
+    cursor = db_v32.conn.cursor()
+    # check that you cannot add blockchain column in xpub_mappings
+    with pytest.raises(sqlcipher.OperationalError) as exc_info:  # pylint: disable=no-member
+        cursor.execute(
+            'INSERT INTO xpub_mappings(address, xpub, derivation_path, account_index, derived_index, blockchain) '  # noqa: 501
+            'VALUES ("1234", "abcd", "d", 3, 6, "BCH");',
+        )
+    assert 'cannot INSERT into generated column "blockchain"' in str(exc_info)
+    xpub_mappings = cursor.execute('SELECT * FROM xpub_mappings').fetchall()
```
Comment (LefterisJP) [inline rotkehlchen/tests/db/test_db_upgrades.py:2625]: nitpick:

Can you also check that the xpub mapping before has the values you inserted in the DB and that they make sense? Make sure the values are real/sensible.

**Review by LefterisJP — COMMENTED** [review]
Hey @prettyirrelevant 

Thanks a lot. Code is looking much better. Left quite a few comments again but most are super easy to address.

Please address them and we can merge.

**Review by LefterisJP — APPROVED** [review]
Perfect. Thank you @prettyirrelevant

**Pull request merged**